### PR TITLE
Write some conventions for dependencies.yaml

### PR DESCRIPTION
**Pull request by KyleFromNVIDIA**
Document the conventions that we are using in `dependencies.yaml` to ensure consistency across RAPIDS repos.

### CODE REVIEW

File: docs/docs/packaging.md [hunk 1]
```diff
@@ -85,3 +85,61 @@ With conda, dependency constraints of the form `X.Y.*` will permit the installat
 pip is stricter and will not generally install nightlies unless the `--pre` argument is specified.
 However, pip can be instructed to install nightlies of specific packages by [using any specifier containing a pre-release or development version](https://pip.pypa.io/en/stable/cli/pip_install/#pre-release-versions).
 Therefore, during wheel builds we also do this [in our CI scripts by adding a `>=0.0.0a0` specifier to all the relevant constraints](https://github.com/rapidsai/cudf/blob/branch-24.06/ci/build_wheel.sh#L34).
+
+## `dependencies.yaml` Conventions
+
+Dependency sets (under the `dependencies` top-level key) shall have the
```
Comment (bdice) [inline docs/docs/packaging.md:91]: Let's replace "dependency sets" with "dependency lists". I think that's how the rapids-dependency-file-generator README names them.

File: docs/docs/packaging.md [hunk 2]
```diff
@@ -85,3 +85,61 @@ With conda, dependency constraints of the form `X.Y.*` will permit the installat
...
+following conventions:
+
+- Build dependencies (under `build-system.requires`) shall be under a
+  dependency set named `py_build_<wheel_name>`.
+- RAPIDS build dependencies (under `tool.rapids-build-backend.requires`) shall
```
Comment (bdice) [inline docs/docs/packaging.md:96]:
```suggestion
- RAPIDS build backend dependencies (under `tool.rapids-build-backend.requires`) shall
```

File: docs/docs/packaging.md [hunk 3]
```diff
@@ -85,3 +85,61 @@ With conda, dependency constraints of the form `X.Y.*` will permit the installat
...
+  followed by a matrix entry with `use_cuda_wheels: "false"` and no packages,
+  to ensure that use cases like DLFW and devcontainers don't get CUDA wheels.
+  The fallback matrix entry shall contain all of the wheels without the `-cu*`
+  suffix. This list will never be used in an actual build, but ensures that
+  the default CUDA wheels that appear in `pyproject.toml` in source control
+  represent a CUDA-version-neutral version of the list of needed CUDA wheels.
+
+    Example:
+
+    ```
```
Comment (bdice) [inline docs/docs/packaging.md:121]:
```suggestion
    ```yaml
```

File: docs/docs/packaging.md [hunk 4]
```diff
@@ -85,3 +85,61 @@ With conda, dependency constraints of the form `X.Y.*` will permit the installat
 pip is stricter and will not generally install nightlies unless the `--pre` argument is specified.
 However, pip can be instructed to install nightlies of specific packages by [using any specifier containing a pre-release or development version](https://pip.pypa.io/en/stable/cli/pip_install/#pre-release-versions).
 Therefore, during wheel builds we also do this [in our CI scripts by adding a `>=0.0.0a0` specifier to all the relevant constraints](https://github.com/rapidsai/cudf/blob/branch-24.06/ci/build_wheel.sh#L34).
+
+## `dependencies.yaml` Conventions
```
Comment (bdice) [inline docs/docs/packaging.md:89]: Other conventions that we ought to document:
- `depends_on_X`
- `--extra-index-url` handling

File: docs/docs/packaging.md (hunk 3 above)
Comment (KyleFromNVIDIA) [reply docs/docs/packaging.md:121]: I tried this but `mkdocs` didn't like it. Is there a plugin that we need?

Comment (bdice) [reply docs/docs/packaging.md:122]: I don't know anything about `mkdocs`. A brief search suggests that "fenced code" is a default plugin. https://mkdocs.readthedocs.io/en/latest/user-guide/configuration/#markdown_extensions

File: docs/docs/packaging.md [hunk 5]
```diff
@@ -85,3 +85,63 @@ With conda, dependency constraints of the form `X.Y.*` will permit the installat
...
+
+Dependency lists (under the `dependencies` top-level key) shall have the
+following conventions:
+
+- Build dependencies (under `build-system.requires`) shall be under a
+  dependency list named `py_build_<wheel_name>`.
+- RAPIDS build backend dependencies (under
+  `tool.rapids-build-backend.requires`) shall be under a dependency list named
+  `py_rapids_build_<wheel_name>`.
```
Comment (jameslamb) [inline docs/docs/packaging.md:98]:
```suggestion
- Build-time dependencies intended for the `build-system.requires` table shall be under a
  dependency list named `py_build_<wheel_name>`. This should only be packages
  that are required for `rapids-build-backend` to run and generate wheel metadata. For example:
    - `rapids-build-backend` itself
    - whatever provides the build backend indicated in `tool.rapids-build-backend.build-backend` (e.g.`scikit-build-core` or `setuptools`)
    - if the project uses a `setup.py`, and libraries that are imported within that `setup.py`
- Build-time dependencies intended for the `tool.rapids-build-backend.requires` table
  shall be under a dependency list named `py_rapids_build_<wheel_name>`. This should be
  packages that are needed to create the files that end up in the wheel. For example:
    - build tools like `ninja` and `cmake`
    - `Cython`
    - packages providing headers / libraries used during compilation (e.g. `librmm`, `libcudf`)
```

I think the phrase "RAPIDS build backend dependencies" here is confusing. I could see that being interpreted as "the things required to run `rapids-build-backend`", and I also think that to someone who isn't already very familiar with this setup, the difference between "RAPIDS build backend dependencies" and "Build dependencies" would not be clear.

Would you consider something like this proposed revision, that separates these two lists by *purpose* and provides some examples?

**Review by jameslamb — COMMENTED** [review]
Thanks for this! Overall I agree with the things you've described. Left one suggestion for clarifying the difference between the two sets of build-time dependencies in `pyproject.toml`.

**Comment by vyasr** [comment]
This is the kind of PR that can easily linger without merging for a long time because there's no specific requirement. There is also no real cost to merging a partial improvement though since it is just a docs fix. @KyleFromNVIDIA can we set a merge deadline and say we'll merge whatever is finished by then? Anything that goes in would be an improvement so it doesn't have to be perfect.

I have not reviewed this yet, but if others think my input would be helpful I'm happy to take a pass.

File: docs/docs/packaging.md (hunk 5 above)
Comment (KyleFromNVIDIA) [reply docs/docs/packaging.md]: Applied your suggestion.

**Comment by KyleFromNVIDIA** [comment]
I've applied @jameslamb's suggestion. I think we can go ahead and merge this.

**Review by jameslamb — APPROVED** [review]
No other comments from me!

I'm happy to do a followup handling these suggestions if you'd like: https://github.com/rapidsai/build-planning/pull/96/files#r1725757113. I've been working with them recently and am familiar with them.

File: docs/docs/packaging.md (hunk 4 above)
Comment (jameslamb) [reply docs/docs/packaging.md:89]: I did this in https://github.com/rapidsai/build-planning/pull/107

**Pull request merged**